src/OpenTelemetry.Instrumentation.EntityFrameworkCore/Implementation/EntityFrameworkDiagnosticListener.cs

```diff
@@ -86,8 +89,7 @@ public override void OnCustom(string name, Activity activity, object payload)
                     }
 
                     var connection = this.connectionFetcher.Fetch(command);
-                    var database = (string)this.databaseFetcher.Fetch(connection);
-
+                    var database = this.options.DisplayNameFunc(this.databaseFetcher, connection);
```

Comment:
We have a standard Enrich pattern followed in the instrumentation libraries Here's ASP .NET instrumentation for reference:
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/AspNetInstrumentationOptions.cs#L53
- https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.AspNet/Implementation/HttpInListener.cs#L112-L119

You could have the option defined as `Action<Activity, string, object> Enrich` and provide the appropriate object when calling `Enrich` for each case. Something like this:
```csharp
switch (name)
        {
            case EntityFrameworkCoreCommandCreated:
                ...
                if (activity.IsAllDataRequested)
                {
                    try
                    {
                        this.options.Enrich?.Invoke(activity, EntityFrameworkCoreCommandCreated, connection); // provide connection object
                    }
                    catch (Exception ex)
                    {
                        EntityFrameworkCoreInstrumentationEventSource.Log.EnrichmentException("EntityFrameworkCoreCommandCreated", ex);
                    }
                }

                break;

            case EntityFrameworkCoreCommandExecuting:
                ...
                if (activity.IsAllDataRequested)
                {
                    try
                    {
                        this.options.Enrich?.Invoke(activity, EntityFrameworkCoreCommandExecuting, command); // provide command object
                    }
                    catch (Exception ex)
                    {
                        EntityFrameworkCoreInstrumentationEventSource.Log.EnrichmentException(EntityFrameworkCoreCommandExecuting, ex);
                    }
                }

                break;

            case EntityFrameworkCoreCommandExecuted:
                ...
                break;

            case EntityFrameworkCoreCommandError:
                ...
                if (activity.IsAllDataRequested)
                {
                    try
                    {
                        this.options.Enrich?.Invoke(activity, EntityFrameworkCoreCommandError, exception); // provide exception
                    }
                    catch (Exception ex)
                    {
                        EntityFrameworkCoreInstrumentationEventSource.Log.EnrichmentException(EntityFrameworkCoreCommandError, ex);
                    }
                }

                break;
        }
```

Reply:
For the purposes of this PR, you would need to call `Enrich` only for `case EntityFrameworkCoreCommandCreated:`.

Reply:
I spoke with @utpilla offline a little about this.

It appears the payload we receive is actually an `IDbCommand` object. The connection is a property of the command. It seems it would simpler to just have an `EnrichWithIDbCommand` delegate invoked upon `EntityFrameworkCoreCommandCreated`. With access to both the command and connection, I think this could serve more use cases for enrich.

Also, it seems like there may be no need to invoke a separate enrich delegate for the other events `EntityFrameworkCoreCommandExecuting` and `EntityFrameworkCoreCommandExecuted` since I think the payload for these too are the same IDbCommand. Exceptions is still an interesting use case, though that could be added separate from this PR.

Reply:
Hi Alan, so playing around with this. I hadn't actually realised the command would be available that way. At CommandCreated level however, the actual command itself isn't available so realistically I think it makes more sense to fully locate it to the executing point. What do you think?

Reply:
Interesting. You mean the CommandText and maybe other properties have not yet been hydrated upon `EntityFrameworkCoreCommandCreated`? In that event, yes I think performing enrich on one of the later events is fine.

Reply:
Correct and cool will do.